akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1324,6 +1328,131 @@ private[stream] object Collect {
     }
 }
 
+object AskStage {
+  final class AskStageTargetActorTerminatedException(ref: ActorRef)
+    extends RuntimeException(s"Target Actor of ask() stage has terminated! Was: ${ref}")
+}
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final case class AskStage[In, Out](targetRef: ActorRef, tag: ClassTag[Out], parallelism: Int, timeout: Timeout)
+  extends GraphStage[FlowShape[In, Out]] with AskSupport { stage ⇒
+  require(timeout ne null, "timeout must not be null")
+  import AskStage._
+
+  private val in = Inlet[In]("Ask.in")
+  private val out = Outlet[Out]("Ask.out")
+
+  override def initialAttributes = DefaultAttributes.ask
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler with StageLogging {
+
+      implicit val t: Timeout = timeout // TODO should be also an attribute?
+      implicit def _tag: ClassTag[Out] = tag
+
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider)
+        .getOrElse(Supervision.stoppingDecider)
+
+      private val futureCB = getAsyncCallback[MapAsync.Holder[Out]](holder ⇒
+        holder.elem match {
+          case Success(succ) ⇒
+            pushNextIfPossible()
+
+          case Failure(ex) ⇒
+            holder.supervisionDirectiveFor(decider, ex) match {
+              // fail fast as if supervision says so
+              case Supervision.Stop ⇒ failStage(ex)
+              case _                ⇒ pushNextIfPossible()
+            }
+        })
+
+      private var buffer: BufferImpl[MapAsync.Holder[Out]] = _
+
+      private lazy val self = getStageActor {
+        case (_, Terminated(`targetRef`)) ⇒
+          failStage(new AskStageTargetActorTerminatedException(targetRef))
+      }
```

Comment:
would it be possible to extend the ordinary `MapAsync` stage instead and add this watch? What else is different?

Reply:
A bit painful... since I want the sender passed into the `ask` to be the `self` of this stage for trackability in timeout messages, I can't get the `self` in the constructor. So it would mean doing a mutable `var` with the MapAsync impl which is set from the preStart of the child... Kind of weird :-\ Should I attempt it? While this is not DRY, it's not so super weird at least

Reply:
I have a similar concern. The mapAsync stage has been buggy and hard to maintain. Now, duplicating will only make it worse.

Do we even need this watch? Won't the timeout be enough?